File: python/l402kit/providers/opennode.py
import time, base64, json
import httpx
from ..types import Invoice


class OpenNodeProvider:
    def __init__(self, api_key: str, test_mode: bool = False):
        self.api_key = api_key
        self.base_url = "https://dev-api.opennode.com" if test_mode else "https://api.opennode.com"

    async def create_invoice(self, amount_sats: int) -> Invoice:
        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self.base_url}/v1/charges",
                headers={"Authorization": self.api_key},
                json={"amount": amount_sats, "description": "L402 API access", "currency": "SATS"},
            )
            res.raise_for_status()
            data = res.json()["data"]
            macaroon = base64.b64encode(json.dumps({"hash": data["id"], "exp": int(time.time()) + 3600}).encode()).decode()
            return Invoice(data["lightning_invoice"]["payreq"], data["id"], macaroon, amount_sats, int(time.time()) + 3600)

    async def check_payment(self, charge_id: str) -> bool:
        async with httpx.AsyncClient() as client:
            res = await client.get(f"{self.base_url}/v1/charge/{charge_id}", headers={"Authorization": self.api_key})
            return res.is_success and res.json().get("data", {}).get("status") == "paid"

File: src/providers/opennode.ts
import type { Invoice, LightningProvider } from "../types";

/**
 * OpenNode — Lightning provider with free sandbox.
 * Get API key: app.opennode.com → API Keys
 */
export class OpenNodeProvider implements LightningProvider {
  private apiKey: string;
  private testMode: boolean;

  constructor(apiKey: string, testMode = false) {
    this.apiKey = apiKey;
    this.testMode = testMode;
  }

  private get baseUrl() {
    return this.testMode ? "https://dev-api.opennode.com" : "https://api.opennode.com";
  }

  async createInvoice(amountSats: number): Promise<Invoice> {
    const res = await fetch(`${this.baseUrl}/v1/charges`, {
      method: "POST",
      headers: { "Content-Type": "application/json", "Authorization": this.apiKey },
      body: JSON.stringify({ amount: amountSats, description: "L402 API access", currency: "SATS", auto_settle: false }),
      signal: AbortSignal.timeout(15_000),
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`OpenNode failed (${res.status}): ${err}`);
    }
    const { data } = (await res.json()) as { data: { lightning_invoice: { payreq: string }; id: string } };
    const macaroon = Buffer.from(JSON.stringify({ hash: data.id, exp: Date.now() + 3_600_000 })).toString("base64");
    return { paymentRequest: data.lightning_invoice.payreq, paymentHash: data.id, macaroon, amountSats, expiresAt: Math.floor(Date.now() / 1000) + 3600 };
  }

  async checkPayment(chargeId: string): Promise<boolean> {
    const res = await fetch(`${this.baseUrl}/v1/charge/${chargeId}`, { headers: { "Authorization": this.apiKey }, signal: AbortSignal.timeout(10_000) });
    if (!res.ok) return false;
    const { data } = (await res.json()) as { data: { status?: string } };
    return data.status === "paid";
  }
}
